BFD-1731 | Cloudwatch Dashboard Import

mariosmith90 opened the pull request:
[BFD-1731](https://jira.cms.gov/browse/BFD-1731)

**User Story or Bug Summary:**

---

### What Does This PR Do?
* Adds to the current codebase the capability to manage Cloudwatch resources. 
* Offers a modular way of iterating upon and making changes to Cloudwatch resources.
* Imports the current AWS Cloudwatch dashboards as code.

### What Should Reviewers Watch For?
* That the following PR enables the management of multiple CloudWatch dashboard. 
* That resources are reused when possible and there is little repetition. 
* The changes represented manage the numerous resources in an easy to follow way. 


### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [ ] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [ ] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [ ] No

* If the answer to any of the questions below is **Yes**, then please add @<!-- -->StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [ ] No

### What Needs to Be Merged and Deployed Before this PR?
This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:
* N/A


### Submitter Checklist
I have gone through and verified that...:

* [ ] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [ ] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [ ] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [ ] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [ ] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [ ] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [ ] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

mjburling commented on ops/terraform/modules/resources/bfd_cw_dashboards/main.tf:
```diff
@@ -0,0 +1,6 @@
+  dashboard_body = templatefile("../../../modules/resources/bfd_cw_dashboards/templates/bfd-dashboards.json", {
+    dashboard_namespace = var.dashboard_namespace
+  })
+}
```
This isn't a blocker, but could you fix your editor to avoid these?

mariosmith90 replied:
> Thanks for the contribution!
> 
> Could you take a look at some of the suggestion? We ought to avoid introducing more variability to the environments with a new module. Let's leverage the environment name from `var.env_config.env` instead.

@mjburling Changes have been made to the branch as requested,

mjburling reviewed (changes requested):
Thanks for the contribution!

Could you take a look at some of the suggestion? We ought to avoid introducing more variability to the environments with a new module. Let's leverage the environment name from `var.env_config.env` instead.

mjburling reviewed (commented):
> @mjburling Changes have been made to the branch as requested,

This solution for the bfd-server dashboards is looking pretty good!

Have you given any thought to adjusting your editor to avoid these _No newline at end of file_ marks in the change set? For instance, the only difference between `ops/terraform/modules/stateful/variables.tf` here and what's in the trunk is the removal of the linefeed. Again, it's not a blocker, but can you address my feedback?

I pulled your changes in and looked over the plans–as expected, it looks like we'll have some well-defined server dashboards via terraform! So far, this is all creating new resources, though. When do you intend on doing the init-import-plan-apply steps?

Additionally, I just reviewed the acceptance criteria for BFD-1731... these changes look to satisfy most of that story, but is that _minimal_ [bfd-pipeline-prod dashboard](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#dashboards:name=bfd-pipeline-prod) already covered by the existing modules in use?

Finally, could you review the PR description and fill out the details from the template and adjust the title?

This looks like we're pretty close. Thanks!

mariosmith90 commented:
> > @mjburling Changes have been made to the branch as requested,
> 
> This solution for the bfd-server dashboards is looking pretty good!
> 
> Have you given any thought to adjusting your editor to avoid these _No newline at end of file_ marks in the change set? For instance, the only difference between `ops/terraform/modules/stateful/variables.tf` here and what's in the trunk is the removal of the linefeed. Again, it's not a blocker, but can you address my feedback?
> 
> I pulled your changes in and looked over the plans–as expected, it looks like we'll have some well-defined server dashboards via terraform! So far, this is all creating new resources, though. When do you intend on doing the init-import-plan-apply steps?
> 
> Additionally, I just reviewed the acceptance criteria for BFD-1731... these changes look to satisfy most of that story, but is that _minimal_ [bfd-pipeline-prod dashboard](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#dashboards:name=bfd-pipeline-prod) already covered by the existing modules in use?
> 
> Finally, could you review the PR description, fill out the details from the template, and adjust the title?
> 
> This looks like we're pretty close. Thanks!

I have added "files.insertFinalNewline": true to my visual studio code.

mariosmith90 commented:
> > @mjburling Changes have been made to the branch as requested,
> 
> This solution for the bfd-server dashboards is looking pretty good!
> 
> Have you given any thought to adjusting your editor to avoid these _No newline at end of file_ marks in the change set? For instance, the only difference between `ops/terraform/modules/stateful/variables.tf` here and what's in the trunk is the removal of the linefeed. Again, it's not a blocker, but can you address my feedback?
> 
> I pulled your changes in and looked over the plans–as expected, it looks like we'll have some well-defined server dashboards via terraform! So far, this is all creating new resources, though. When do you intend on doing the init-import-plan-apply steps?
> 
> Additionally, I just reviewed the acceptance criteria for BFD-1731... these changes look to satisfy most of that story, but is that _minimal_ [bfd-pipeline-prod dashboard](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#dashboards:name=bfd-pipeline-prod) already covered by the existing modules in use?
> 
> Finally, could you review the PR description, fill out the details from the template, and adjust the title?
> 
> This looks like we're pretty close. Thanks!

I have added "files.insertFinalNewline": true to my visual studio code.

> > @mjburling Changes have been made to the branch as requested,
> 
> This solution for the bfd-server dashboards is looking pretty good!
> 
> Have you given any thought to adjusting your editor to avoid these _No newline at end of file_ marks in the change set? For instance, the only difference between `ops/terraform/modules/stateful/variables.tf` here and what's in the trunk is the removal of the linefeed. Again, it's not a blocker, but can you address my feedback?
> 
> I pulled your changes in and looked over the plans–as expected, it looks like we'll have some well-defined server dashboards via terraform! So far, this is all creating new resources, though. When do you intend on doing the init-import-plan-apply steps?
> 
> Additionally, I just reviewed the acceptance criteria for BFD-1731... these changes look to satisfy most of that story, but is that _minimal_ [bfd-pipeline-prod dashboard](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#dashboards:name=bfd-pipeline-prod) already covered by the existing modules in use?
> 
> Finally, could you review the PR description, fill out the details from the template, and adjust the title?
> 
> This looks like we're pretty close. Thanks!

I have added "files.insertFinalNewline": true

It's all creating new resources since terraform import requires an individual resource block for each imported AWS resource. This is not ideal for our current modular code layout. As for the terraform workflow mentioned above, the intention is to do this once it has been peer approved. The idea was to rename the previous dashboards and deploy these in their place until the old ones are deleted at an agreed upon time.

As for the bad pipeline, I couldn't think of a logical way at face value to implement them in to the current workflow. Perhaps they can be added to the mgmt modul, as they don't fit the prod/prod-sbx/test paradigm that we're currently using.

mjburling commented:
I wanted to get back to this sooner–I apologize for taking much of the day to get back to this.

> I have added "files.insertFinalNewline": true

Oh, good. Can you go ahead and rewrite those files to drop the EOF character warning? 

> It's all creating new resources since terraform import requires an individual resource block for each imported AWS resource. This is not ideal for our current modular code layout.

I don't think I follow this fully, but I could always be missing something. Where resources are defined, whether in the root or other referenced module, shouldn't have any impact on this behavior. See below.

> As for the terraform workflow mentioned above, the intention is to do this once it has been peer approved. The idea was to rename the previous dashboards and deploy these in their place until the old ones are deleted at an agreed upon time.

This import work can get to be messy, but it doesn't have to! I just tried running some evaluations, and I'm getting some errors with how this is currently implemented. We might want to pair through this.

> As for the bfd pipeline, I couldn't think of a logical way at face value to implement them in to the current workflow. Perhaps they can be added to the mgmt module, as they don't fit the prod/prod-sbx/test paradigm that we're currently using.

But it does follow the same paradigm. It's just that the dashboards for pipeline in the test and prod-sbx environments don't exist and won't need to be imported. Let's address that in a separate PR in the interest of shipping the changes to support the bfd-server.

mjburling commented:
Revised and resolved in #1132
